File: src/main/java/com/dsm/quiz/model/user/service/AuthService.java
package com.dsm.quiz.model.user.service;

import com.dsm.quiz.exception.UserAlreadyException;
import com.dsm.quiz.exception.UserNicknameAlreadyException;
import com.dsm.quiz.model.user.User;
import com.dsm.quiz.model.user.dto.UserCreateRequestDto;
import com.dsm.quiz.model.user.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.security.crypto.password.AbstractPasswordEncoder;
import org.springframework.stereotype.Service;

@Service
@RequiredArgsConstructor
public class AuthService {
    private final UserRepository userRepository;
    private final AbstractPasswordEncoder passwordEncoder;
    public void join(UserCreateRequestDto user){
        if (userRepository.existsById(user.getEmail())){
            throw new UserAlreadyException();
        }
        if (userRepository.existsByNickname(user.getNickname())){
            throw new UserNicknameAlreadyException();
        }

        userRepository.save(
                User.builder()
                .email(user.getEmail())
                .password(passwordEncoder.encode(user.getPassword()))
                .nickname(user.getNickname()).build()
        );

    }
}

File: src/main/kotlin/com/quiz/qq/domain/user/service/AuthService.kt
package com.quiz.qq.domain.user.service

import com.quiz.qq.domain.user.User
import com.quiz.qq.domain.user.UserRepository
import com.quiz.qq.domain.user.dto.UserCreateRequestDto
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.stereotype.Service

@Service
class AuthService(private val repository: UserRepository, private val passwordEncoder: PasswordEncoder) {
    fun join(dto: UserCreateRequestDto){
        repository.save(
            User(dto.id,
                passwordEncoder.encode(dto.password),
                dto.nickname)
        )
    }
}